### PR TITLE
feat: auto compound staking

### LINKED ISSUE #960: Staking Auto compounded feature
Moonbeam has published a pallet for auto-staking feature.
We should have it too.

### PR DESCRIPTION
**Pull request by wangminqi**
Still need Migration.
And benchmark #977 
This closes #960

### CODE REVIEW

**Comment by wangminqi** [comment]
migration added

File: pallets/parachain-staking/src/lib.rs [hunk 1]
```diff
@@ -1551,19 +1598,36 @@ pub mod pallet {
 				collator_count = collator_count.saturating_add(1u32);
 				delegation_count = delegation_count.saturating_add(state.delegation_count);
 				total = total.saturating_add(state.total_counted);
-				let snapshot_total = state.total_counted;
-				let top_rewardable_delegations = Self::get_rewardable_delegators(account);
+				let CountedDelegations { uncounted_stake, rewardable_delegations } =
```
Comment (wangminqi) [inline pallets/parachain-staking/src/lib.rs:1601]: remark

File: pallets/parachain-staking/src/auto_compound.rs [hunk 2]
```diff
@@ -0,0 +1,319 @@
+// Copyright 2019-2022 PureStake Inc.
```
Comment (Kailai-Wang) [inline pallets/parachain-staking/src/auto_compound.rs]: licence

File: pallets/parachain-staking/src/lib.rs (hunk 1 above)
Comment (wangminqi) [reply pallets/parachain-staking/src/lib.rs:1601]: https://github.com/PureStake/moonbeam/commit/3f41d5481d502a59437c11be735b4bd05cc6781e

File: pallets/parachain-staking/src/auto_compound.rs (hunk 2 above)
Comment (wangminqi) [reply pallets/parachain-staking/src/auto_compound.rs]: done!

File: runtime/rococo/src/migration/mod_backup.rs [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+// (at your option) any later version.
+//
+// Litentry is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Litentry.  If not, see <https://www.gnu.org/licenses/>.
+#![allow(deprecated)]
```
Comment (jingleizhang) [inline runtime/rococo/src/migration/mod_backup.rs:16]: Let's delete this file when the migration is completed.

File: pallets/parachain-staking/src/auto_compound.rs [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Litentry.  If not, see <https://www.gnu.org/licenses/>.
+
+//! Inspired by:
+//! - Moonbeam `pallet_parachain_staking`
+//! implementations.
+
+//! Auto-compounding functionality for staking rewards
```
Comment (DogLi) [inline pallets/parachain-staking/src/auto_compound.rs:21]: I think Moonbeam's CopyRight should be added here.

```
//! adapted from https://github.com/PureStake/moonbeam/blob/master/pallets/parachain-staking/src/auto_compound.rs
//! Copyright 2019-2022 PureStake Inc.(licensed under the GNU General Public License)
//! Modifications Copyright (c) 2020-2022 Litentry Technologies GmbH. (licensed under GNU General Public License)
```

**Review by jingleizhang — APPROVED** [review]
tested `setAutoCompound`, `delegatorBondMore`, `delegateWithAutoCompound` and `atStake` on the roccoco network (wss://rpc.rococo-parachain-sg.litentry.io)

**Pull request merged**